library/std/src/sys/windows/compat.rs

```diff
@@ -49,81 +49,211 @@
 //! * call any Rust function or CRT function that touches any static
 //!   (global) state.
 
-macro_rules! compat_fn {
-    ($module:literal: $(
+use crate::ffi::{c_void, CStr};
+use crate::ptr::NonNull;
+use crate::sys::c;
+
+/// Helper macro for creating CStrs from literals and symbol names.
+macro_rules! ansi_str {
+    (sym $ident:ident) => {{
+        #[allow(unused_unsafe)]
+        unsafe {
+            CStr::from_bytes_with_nul_unchecked(concat!(stringify!($ident), "\0").as_bytes())
+        }
+    }};
+    ($lit:literal) => {{
+        #[allow(unused_unsafe)]
+        unsafe {
+            CStr::from_bytes_with_nul_unchecked(concat!($lit, "\0").as_bytes())
+        }
+    }};
+}
+
+#[used]
+#[link_section = ".CRT$XCU"]
+static INIT_TABLE_ENTRY: unsafe extern "C" fn() = init;
+
+/// This is where the magic preloading of symbols happens.
+///
+/// We don't preload all lazily loaded functions, only those which don't have
+/// a fallback or may be particularly perf sensitive.
+unsafe extern "C" fn init() {
+    // There is no locking here. This code is executed before main() is entered, and
+    // is guaranteed to be single-threaded.
+    //
+    // DO NOT do anything interesting or complicated in this function! DO NOT call
+    // any Rust functions or CRT functions if those functions touch any global state,
+    // because this function runs during global initialization. For example, DO NOT
+    // do any dynamic allocation, don't call LoadLibrary, etc.
+
+    if let Some(synch) = Module::new(c::SYNCH_API) {
+        // These are optional and so we must manually attempt to load them
+        // before they can be used.
+        c::WaitOnAddress::preload(synch);
+        c::WakeByAddressSingle::preload(synch);
+    }
+
+    if let Some(kernel32) = Module::new(c::KERNEL32) {
+        // Preloading this means getting a precise time will be as fast as possible.
+        c::GetSystemTimePreciseAsFileTime::preload(kernel32);
+    }
+
+    if let Some(ntdll) = Module::new(c::NTDLL) {
+        // A program is almost certainly going to be doing some kind of I/O.
+        c::NtReadFile::preload(ntdll);
+        c::NtWriteFile::preload(ntdll);
+        c::RtlNtStatusToDosError::preload(ntdll);
+    }
```

Comment:
#81478 introduced this pre-loading but is a little vague on specifics. Tbh, I'm not terribly fussed about this one way or the other.

Basically the only advantage of pre-loading is that it avoids some small perf unpredictability when calling certain functions for the first time. The disadvantage of pre-loading is it's slightly more work on startup and forces symbols to be referenced and loaded even if they're never used by anything.

My only thinking with those three highlighted functions were that there was no reason why not as they're almost certainly are used (e.g. panics use `NtWriteFile` and `RtlNtStatusToDosError`). But that's admittedly a questionable decision because I doubt there's any perf justification in the context of performing I/O.

Reply:
I've now added a note which clarifies my thinking, and therefore removed those three functions based on that reasoning.